streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java

```diff
@@ -74,20 +198,220 @@ public static Map<ProcessId, KafkaStreamsAssignment> optimizeRackAwareStandbyTas
+        public final Graph<Integer> graph;
+        public final Map<TaskId, UUID> clientByTask;
+        public final Map<UUID, Integer> taskCountByClient;
+
+        public AssignmentGraph(final Graph<Integer> graph,
+                               final Map<TaskId, UUID> clientByTask,
+                               final Map<UUID, Integer> taskCountByClient) {
+            this.graph = graph;
+            this.clientByTask = clientByTask;
+            this.taskCountByClient = taskCountByClient;
+        }
+    }
+
     /**
-     * Return a "no-op" assignment that just copies the previous assignment of tasks to KafkaStreams clients
      *
-     * @param applicationState the metadata and other info describing the current application state
+     * @return the traffic cost of assigning this {@param task} to the client {@param streamsState}.
+     */
+    private static int getCrossRackTrafficCost(final Set<TaskTopicPartition> topicPartitions,
+                                               final Optional<String> clientRack,
+                                               final int crossRackTrafficCost) {
+        if (!clientRack.isPresent()) {
+            throw new IllegalStateException("Client doesn't have rack configured.");
+        }
+
+        int cost = 0;
+        for (final TaskTopicPartition topicPartition : topicPartitions) {
+            final Optional<Set<String>> topicPartitionRacks = topicPartition.rackIds();
+            if (topicPartitionRacks == null || !topicPartitionRacks.isPresent()) {
+                throw new IllegalStateException("TopicPartition " + topicPartition + " has no rack information.");
+            }
+
+            if (topicPartitionRacks.get().contains(clientRack.get())) {
+                continue;
+            }
+
+            cost += crossRackTrafficCost;
+        }
+        return cost;
+    }
+
+    /**
+     * This function returns whether the current application state has the required rack information
+     * to make assignment decisions with.
+     * This includes ensuring that every client has a known rack id, and that every topic partition for
+     * every logical task that needs to be assigned also has known rack ids.
+     * If a logical task has no source topic partitions, it will return false.
+     * If standby tasks are configured, but a logical task has no changelog topic partitions, it will return false.
      *
-     * @return a new map containing an assignment that replicates exactly the previous assignment reported
-     *         in the applicationState
+     * @return whether rack-aware assignment decisions can be made for this application.
      */
-    public static Map<ProcessId, KafkaStreamsAssignment> identityAssignment(
-        final ApplicationState applicationState
-    ) {
-        throw new UnsupportedOperationException("Not Implemented.");
+    public static boolean hasValidRackInformation(final ApplicationState applicationState) {
```

Comment:
needs to be private, otherwise it would be a public API which we don't want (and we don't need IIUC)